In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

In [3]:
# Input preperation
_data = [   [10,18],
            [7,18],[4,17],[3,16],[1,13],
            [1,10],[11,10],[16,16],[21,6],
            [26,10],[10,1],[13,7],[18,12],
            [22,18],[23,10],[8,9],[6,6],
            [9,8],[13,5],[22,6],[5,1],
            [19,5],[4,9],[6,9],[3,5],
            [3,2],[11,3],[19,10],[26,17],
            [21,4],[22,6],[20,7],[18,16],
            [8,3],[21,13],[8,15],[6,14],
            [9,4],[25,14],[7,4],[9,14],
            [22,14],[5,13],[8,2],[6,3],
            [7,3],[8,14],[22,15],[9,12],
            [7,12]
        ]
# Output [Blue, Red]
_target = [[1,0],[1,0],[1,0],[1,0]
          ,[1,0],[1,0],[1,0],[1,0]
          ,[1,0],[1,0],[1,0],[1,0]
          ,[1,0],[1,0],[1,0],[1,0]
          ,[1,0],[1,0],[1,0],[1,0]
          ,[1,0],[1,0],[1,0],[1,0]
          ,[1,0],[1,0],[1,0],[1,0]
          ,[1,0],[1,0],[1,0],[1,0]
          ,[1,0]
          ,[0,1],[0,1],[0,1],[0,1]
          ,[0,1],[0,1],[0,1],[0,1]
          ,[0,1],[0,1],[0,1],[0,1]
          ,[0,1],[0,1],[0,1],[0,1]
          ,[0,1]]

train_X = pd.DataFrame(np.array(_data), columns=['x1','x2'])
train_Y = pd.DataFrame(np.array(_target), columns=['Blue','Red'])

In [4]:
from keras.models import Sequential
from keras.layers import Dense

# create model
model = Sequential()
n_inputs = train_X.shape[1]

# add model layers
model.add(Dense(9,activation='softmax',input_shape=(n_inputs,)))
model.add(Dense(3, activation='softmax'))
model.add(Dense(2))

Using TensorFlow backend.


In [5]:
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

In [6]:
from keras.callbacks import EarlyStopping

In [7]:
es = EarlyStopping(patience=5)
# Train the model
model.fit(train_X, train_Y, validation_split=0.2, epochs=150, callbacks=[es])

Train on 40 samples, validate on 10 samples
Epoch 1/150
40/40 [==============================] - 1s 20ms/step - loss: 0.9546 - accuracy: 0.8250 - val_loss: 0.4449 - val_accuracy: 0.0000e+00
Epoch 2/150
40/40 [==============================] - 0s 268us/step - loss: 0.9293 - accuracy: 0.8250 - val_loss: 0.4653 - val_accuracy: 0.0000e+00
Epoch 3/150
40/40 [==============================] - 0s 242us/step - loss: 0.9007 - accuracy: 0.8250 - val_loss: 0.4852 - val_accuracy: 0.0000e+00
Epoch 4/150
40/40 [==============================] - 0s 227us/step - loss: 0.8749 - accuracy: 0.8250 - val_loss: 0.5052 - val_accuracy: 0.0000e+00
Epoch 5/150
40/40 [==============================] - 0s 222us/step - loss: 0.8511 - accuracy: 0.8250 - val_loss: 0.5255 - val_accuracy: 0.0000e+00
Epoch 6/150
40/40 [==============================] - 0s 234us/step - loss: 0.8281 - accuracy: 0.8250 - val_loss: 0.5460 - val_accuracy: 0.0000e+00


In [8]:
model.get_weights()
#test_pred = model.predict(test_X)

[array([[ 0.22415464,  0.09937146, -0.02442451,  0.13824017, -0.23531228,
         -0.23015037,  0.59779274, -0.1701167 ,  0.5945799 ],
        [ 0.03438755, -0.4697363 ,  0.36498266, -0.09256455,  0.6428428 ,
         -0.03536626, -0.45127004,  0.00660057, -0.37589857]],
       dtype=float32),
 array([-0.00413427,  0.0093336 , -0.0091468 , -0.01037611,  0.00945695,
        -0.01031078, -0.01043156, -0.01062863,  0.0105148 ], dtype=float32),
 array([[ 0.4376258 , -0.32763508,  0.53592575],
        [-0.52530235,  0.1589784 ,  0.66948694],
        [-0.22961761,  0.5568017 ,  0.10889921],
        [ 0.55441666, -0.06636168,  0.3343891 ],
        [ 0.28172737,  0.29323488,  0.51443905],
        [ 0.6603879 , -0.39133945, -0.45147234],
        [-0.41493815, -0.05143942, -0.4809096 ],
        [ 0.28295925,  0.25381202, -0.5546837 ],
        [-0.15640965, -0.44522372,  0.48926815]], dtype=float32),
 array([-0.01126225, -0.01125421,  0.01126108], dtype=float32),
 array([[ 0.67680347, -0.5905258

# Answer2 : 

In 3 epochs we attain the accuracy of 82.50% without any further significant improvement.

There are 27 + 30 + 8 trainable parameters.

I have used 'sigmoid' activation function.

In [9]:
# from sklearn.datasets import make_blobs
# Xnew, _ = make_blobs(n_samples=3, centers=2, n_features=2, random_state=1)
from numpy import array
Xnew = array([[15,15],[7,3]])

# Answer4 : Predicting

In [10]:
print('Predict the classes: ')
ynew = model.predict_classes(Xnew)
for i in range(len(Xnew)):
	print("X=%s, Predicted=%s" % (Xnew[i], ynew[i]))

Predict the classes: 
X=[15 15], Predicted=0
X=[7 3], Predicted=0


In [11]:
model.summary()

Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_1 (Dense)              (None, 9)                 27        
_________________________________________________________________
dense_2 (Dense)              (None, 3)                 30        
_________________________________________________________________
dense_3 (Dense)              (None, 2)                 8         
Total params: 65
Trainable params: 65
Non-trainable params: 0
_________________________________________________________________


# Answer3 : Plot

In [12]:
def plot_decision_boundary(X, y, model, steps=1000, cmap='Paired'):
    cmap = plt.get_cmap(cmap)

    # Define region of interest by data limits
    xmin, xmax = X[:,0].min() - 1, X[:,0].max() + 1
    ymin, ymax = X[:,1].min() - 1, X[:,1].max() + 1
    steps = 1000
    x_span = np.linspace(xmin, xmax, steps)
    y_span = np.linspace(ymin, ymax, steps)
    xx, yy = np.meshgrid(x_span, y_span)

    # Make predictions across region of interest
    labels = model.predict(np.c_[xx.ravel(), yy.ravel()])

    # Plot decision boundary in region of interest
    z = labels.reshape(xx.shape)

    fig, ax = plt.subplots()
    ax.contourf(xx, yy, z, cmap=cmap, alpha=0.5)

    # Get predicted labels on training data and plot
    train_labels = model.predict(X)
    ax.scatter(X[:,0], X[:,1], c=y, cmap=cmap, lw=0)

    return fig, ax

In [13]:
train_Y.shape

(50, 2)

In [14]:
plot_decision_boundary(train_X.values, train_Y.values, model)

ValueError: cannot reshape array of size 2000000 into shape (1000,1000)

In [15]:
plt.scatter(train_X[:,0], train_X[:,1])

TypeError: '(slice(None, None, None), 0)' is an invalid key